### PR TITLE
Add reading of public key from buffer 

### PR DESCRIPTION
**Pull request by ghost**
# What this PR does?
* Add reading of public key from buffer 

# How to build:

## Linux / Mac

* `git clone -b feature/key-from-str git@github.com:RealImage/jwtxx.git`
* `cd jwtxx/build`
* `cmake  .. `
* `make -j `

# How to test:
* `make tests`
* `make test`

### CODE REVIEW

**Review by yuriyvolkov — APPROVED** [review]

File: src/utils.cpp [hunk 1]
```diff
@@ -42,12 +42,25 @@ std::string sysError() noexcept
...
-    return Utils::EVPKeyPtr(PEM_read_PUBKEY(fp.get(), nullptr, nullptr, nullptr));
+    FilePtr fp(fopen(src.c_str(), "rb"));
+
+    // src is file name
+    if (fp)
+        return Utils::EVPKeyPtr(PEM_read_PUBKEY(fp.get(), nullptr, nullptr, nullptr));
+
+    // src is key data
+    unsigned char *decodedData = new unsigned char [src.size()];
+    EVP_DecodeBlock(decodedData, reinterpret_cast<const unsigned char*>(src.data()), static_cast<int>(src.size()));
```
Comment (sivachandran) [inline src/utils.cpp:55]: we can use PEM_read_bio_PUBKEY instead of decoding the base64 by ourself.

File: src/utils.cpp [hunk 2]
```diff
@@ -42,12 +42,25 @@ std::string sysError() noexcept
...
 }
 
-Utils::EVPKeyPtr readPublicKey(const std::string& fileName)
+Utils::EVPKeyPtr readPublicKey(const std::string& src)
 {
-    FilePtr fp(fopen(fileName.c_str(), "rb"));
-    if (!fp)
-        throw JWTXX::Key::Error("Can't open key file '" + fileName + "'. " + sysError());
-    return Utils::EVPKeyPtr(PEM_read_PUBKEY(fp.get(), nullptr, nullptr, nullptr));
+    FilePtr fp(fopen(src.c_str(), "rb"));
```
Comment (sivachandran) [inline src/utils.cpp:47]: we could check whether the string starts with `-----BEGIN` and if so then we can straightway try to read pubkey from the base64. If it doesn't start with then we can try to open it as file.

**Review by sivachandran — CHANGES_REQUESTED** [review]

File: src/utils.cpp (hunk 1 above)
Comment (ghost) [reply src/utils.cpp:55]: @sivachandran  We have to pass string in PEM format if we use `PEM_read_bio_PUBKEY`. This code allows us to use a string without any formatting.
See discussion here
http://www.archivum.info/openssl-users@openssl.org/2010-05/00092/RE-PEM_read_bio_PUBKEY-does-not-do-the-trick.html

File: src/utils.cpp (hunk 2 above)
Comment (ghost) [reply src/utils.cpp:47]: @sivachandran It is not our case, we pass raw string without `BEGIN/END`
See test https://github.com/RealImage/jwtxx/pull/1/files#diff-c221b935bd4506307c4782e4f01b441eR26

File: src/utils.cpp (hunk 1 above)
Comment (sivachandran) [reply src/utils.cpp:55]: @AntonQubecinema I'm recommending we pass with `BEGIN` and `END` marker. It just our use case to keep the public key without these marker. But if you look at the general use case the public key would always carry these markers.

Comment (sivachandran) [reply src/utils.cpp:55]: @AntonQubecinema As we are going to contribute this functionality it is better to keep the input format generic. In html2dcp we can construct the PEM public key from the modulus and exponent we have.

**Pull request closed without merging**